src/backend/distributed/connection/shared_connection_stats.c

```diff
@@ -0,0 +1,617 @@
+static void LockConnectionSharedMemory(LWLockMode lockMode);
+static void UnLockConnectionSharedMemory(void);
+static void SharedConnectionStatsShmemInit(void);
+static size_t SharedConnectionStatsShmemSize(void);
+static int SharedConnectionHashCompare(const void *a, const void *b, Size keysize);
+static uint32 SharedConnectionHashHash(const void *key, Size keysize);
+
+
+PG_FUNCTION_INFO_V1(citus_remote_connection_stats);
+
+
+/*
+ * citus_remote_connection_stats returns all the avaliable information about all
+ * the remote connections (a.k.a., connections to remote nodes).
+ */
+Datum
+citus_remote_connection_stats(PG_FUNCTION_ARGS)
+{
+	TupleDesc tupleDescriptor = NULL;
+
+	CheckCitusVersion(ERROR);
+	Tuplestorestate *tupleStore = SetupTuplestore(fcinfo, &tupleDescriptor);
+
+	StoreAllConnections(tupleStore, tupleDescriptor);
+
+	/* clean up and return the tuplestore */
+	tuplestore_donestoring(tupleStore);
+
+	PG_RETURN_VOID();
+}
+
+
+/*
+ * StoreAllConnections gets connections established from the current node
+ * and inserts them into the given tuplestore.
+ *
+ * We don't need to enforce any access privileges as the number of backends
+ * on any node is already visible on pg_stat_activity to all users.
+ */
+static void
+StoreAllConnections(Tuplestorestate *tupleStore, TupleDesc tupleDescriptor)
+{
+	Datum values[REMOTE_CONNECTION_STATS_COLUMNS];
+	bool isNulls[REMOTE_CONNECTION_STATS_COLUMNS];
+
+	/* we're reading all shared connections, prevent any changes */
+	LockConnectionSharedMemory(LW_SHARED);
+
+	HASH_SEQ_STATUS status;
+	SharedConnStatsHashEntry *connectionEntry = NULL;
+
+	hash_seq_init(&status, SharedConnStatsHash);
+	while ((connectionEntry = (SharedConnStatsHashEntry *) hash_seq_search(&status)) != 0)
+	{
+		/* get ready for the next tuple */
+		memset(values, 0, sizeof(values));
+		memset(isNulls, false, sizeof(isNulls));
+
+		char *databaseName = get_database_name(connectionEntry->key.databaseOid);
+		if (databaseName == NULL)
+		{
+			/* database might have been dropped */
+			continue;
+		}
+
+		values[0] = PointerGetDatum(cstring_to_text(connectionEntry->key.hostname));
+		values[1] = Int32GetDatum(connectionEntry->key.port);
+		values[2] = PointerGetDatum(cstring_to_text(databaseName));
+		values[3] = Int32GetDatum(connectionEntry->connectionCount);
+
+		tuplestore_putvalues(tupleStore, tupleDescriptor, values, isNulls);
+	}
+
+	UnLockConnectionSharedMemory();
+}
+
+
+/*
+ * RemoveInactiveNodesFromSharedConnections goes over the SharedConnStatsHash
+ * and removes the inactive entries.
+ */
+void
+RemoveInactiveNodesFromSharedConnections(void)
+{
+	/* we're reading all shared connections, prevent any changes */
+	LockConnectionSharedMemory(LW_EXCLUSIVE);
+
+	HASH_SEQ_STATUS status;
+	SharedConnStatsHashEntry *connectionEntry = NULL;
+
+	int entryCount = hash_get_num_entries(SharedConnStatsHash);
+	if (entryCount + 1 < MaxWorkerNodesTracked)
```

Comment:
@marcocitus This line effectively prevents removing any entries during the regression tests. If you remove this line, you'd see that we hit `Assert (connectionEntry->connectionCount > 0);`.

I'm looking into this, any feedback is welcome on how to improve this.

Reply:
See 1be411f36b7b1368a8a70e81bc8ddef78aa3c916, that seems like a pretty decent approximation to not get into a case where we increment/decrement non-exiting nodes.

I think what I observe is mostly a regression test issue, in a real world use-pattern, we're unlikely to get into the second loop below

Reply:
are we guaranteed that the hash table can take another entry if entryCount < MaxWorkerNodesTracked? Then this seems fine.

Reply:
There is no documentation about that, but looking into the places that this function is used in Postgres (and reading through its implementation), this is implied that there are exactly `MaxWorkerNodesTracked ` spaces that we can use.

Though, shared memory hash tables are little different than other.

I'll spend some more time on this to make sure.

Reply:
Reading through the code, yes we are guaranteed that hash table can take another entry. Because, it is guaranteed that `MaxWorkerNodesTracked ` entries are allocated via `ShmemInitHash`, and `hash_get_num_entries ` returns the exact number of entries in the hash table. Finally, as long as there is space in the hash map, it'll give the entry -- even if all entries collide. 

( Changing `MaxWorkerNodesTracked` requires restart.)